File: src/library_skills/workspace.py
import fnmatch
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import cast

if sys.version_info >= (3, 11):  # pragma: no cover
    import tomllib
else:
    import tomli as tomllib


@dataclass(frozen=True)
class UvWorkspace:
    root: Path
    members: tuple[Path, ...]
    current_member: Path | None


def find_uv_workspace(cwd: Path) -> UvWorkspace | None:
    """Find uv workspace metadata for cwd, if any."""
    for directory in [cwd, *cwd.parents]:
        pyproject = directory / "pyproject.toml"
        if not pyproject.is_file():
            continue
        data = _read_pyproject(pyproject)
        if not _has_uv_workspace(data):
            continue
        members = _find_workspace_members(directory, data)
        return UvWorkspace(
            root=directory,
            members=tuple(members),
            current_member=_find_current_member(cwd, members),
        )
    return None


def workspace_dependency_files(workspace: UvWorkspace) -> list[Path]:
    """Return pyproject files used for default dependency filtering."""
    if workspace.current_member is not None:
        return [workspace.current_member / "pyproject.toml"]
    return [
        path
        for path in [workspace.root / "pyproject.toml"]
        + [member / "pyproject.toml" for member in workspace.members]
        if path.is_file()
    ]


def _read_pyproject(path: Path) -> dict[str, object]:
    try:
        data = tomllib.loads(path.read_text(encoding="utf-8"))
    except (OSError, tomllib.TOMLDecodeError):
        return {}
    return data if isinstance(data, dict) else {}


def _has_uv_workspace(data: dict[str, object]) -> bool:
    tool = data.get("tool")
    if not isinstance(tool, dict):
        return False
    tool_data = cast("dict[str, object]", tool)
    uv = tool_data.get("uv")
    if not isinstance(uv, dict):
        return False
    uv_data = cast("dict[str, object]", uv)
    return isinstance(uv_data.get("workspace"), dict)


def _find_workspace_members(root: Path, data: dict[str, object]) -> list[Path]:
    workspace = _get_workspace_table(data)
    if workspace is None:
        return []
    member_globs = workspace.get("members")
    if not isinstance(member_globs, list):
        return []
    exclude_globs = workspace.get("exclude")
    excludes = (
        [item for item in exclude_globs if isinstance(item, str)]
        if isinstance(exclude_globs, list)
        else []
    )

    members: set[Path] = set()
    for member_glob in member_globs:
        if not isinstance(member_glob, str):
            continue
        for member in root.glob(member_glob):
            if not member.is_dir():
                continue
            relative = member.relative_to(root).as_posix()
            if _is_excluded(relative, excludes):
                continue
            if (member / "pyproject.toml").is_file():
                members.add(member.resolve())
    return sorted(members)


def _get_workspace_table(data: dict[str, object]) -> dict[str, object] | None:
    tool = data.get("tool")
    if not isinstance(tool, dict):
        return None
    tool_data = cast("dict[str, object]", tool)
    uv = tool_data.get("uv")
    if not isinstance(uv, dict):
        return None
    uv_data = cast("dict[str, object]", uv)
    workspace = uv_data.get("workspace")
    return cast("dict[str, object]", workspace) if isinstance(workspace, dict) else None


def _is_excluded(relative_path: str, exclude_globs: list[str]) -> bool:
    return any(
        fnmatch.fnmatch(relative_path, pattern.rstrip("/"))
        or fnmatch.fnmatch(f"{relative_path}/", pattern)
        for pattern in exclude_globs
    )


def _find_current_member(cwd: Path, members: list[Path]) -> Path | None:
    cwd = cwd.resolve()
    matching_members = [
        member for member in members if _is_relative_to(cwd, member.resolve())
    ]
    if not matching_members:
        return None
    return max(matching_members, key=lambda path: len(path.parts))


def _is_relative_to(path: Path, parent: Path) -> bool:
    try:
        path.relative_to(parent)
    except ValueError:
        return False
    return True

File: ts/src/workspace.ts
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { relative, resolve, sep } from "node:path";
import { parse, type TomlTable } from "smol-toml";

export interface UvWorkspace {
	root: string;
	members: string[];
	currentMember: string | null;
}

export function findUvWorkspace(cwd: string): UvWorkspace | null {
	for (const directory of ancestors(cwd)) {
		const pyproject = `${directory}/pyproject.toml`;
		if (!existsSync(pyproject)) {
			continue;
		}
		const data = readPyproject(pyproject);
		if (!hasUvWorkspace(data)) {
			continue;
		}
		const members = findWorkspaceMembers(directory, data);
		return {
			root: directory,
			members,
			currentMember: findCurrentMember(cwd, members),
		};
	}
	return null;
}

export function workspaceDependencyFiles(workspace: UvWorkspace): string[] {
	if (workspace.currentMember !== null) {
		return [`${workspace.currentMember}/pyproject.toml`];
	}
	return [
		`${workspace.root}/pyproject.toml`,
		...workspace.members.map((member) => `${member}/pyproject.toml`),
	].filter((path) => existsSync(path));
}

function readPyproject(path: string): TomlTable {
	try {
		const data = parse(readFileSync(path, "utf8"));
		return isRecord(data) ? data : {};
	} catch {
		/* v8 ignore next -- covers invalid TOML or a race while reading. */
		return {};
	}
}

function hasUvWorkspace(data: TomlTable): boolean {
	const tool = data["tool"];
	if (!isRecord(tool)) {
		return false;
	}
	const uv = tool["uv"];
	if (!isRecord(uv)) {
		return false;
	}
	return isRecord(uv["workspace"]);
}

function findWorkspaceMembers(root: string, data: TomlTable): string[] {
	const workspace = getWorkspaceTable(data);
	if (workspace === null || !Array.isArray(workspace["members"])) {
		return [];
	}
	const excludes = Array.isArray(workspace["exclude"])
		? workspace["exclude"].filter(
				(value): value is string => typeof value === "string",
			)
		: [];
	const members = new Set<string>();
	for (const memberGlob of workspace["members"]) {
		if (typeof memberGlob !== "string") {
			continue;
		}
		for (const member of expandMemberGlob(root, memberGlob)) {
			const relativePath = relative(root, member).split(sep).join("/");
			if (isExcluded(relativePath, excludes)) {
				continue;
			}
			if (isFile(`${member}/pyproject.toml`)) {
				members.add(resolve(member));
			}
		}
	}
	return [...members].sort();
}

function getWorkspaceTable(data: TomlTable): Record<string, unknown> | null {
	const tool = data["tool"];
	if (!isRecord(tool)) {
		/* v8 ignore next -- guarded by hasUvWorkspace before this function is called. */
		return null;
	}
	const uv = tool["uv"];
	if (!isRecord(uv)) {
		/* v8 ignore next -- guarded by hasUvWorkspace before this function is called. */
		return null;
	}
	const workspace = uv["workspace"];
	return isRecord(workspace) ? workspace : null;
}

function expandMemberGlob(root: string, pattern: string): string[] {
	const parts = pattern.split("/");
	const results: string[] = [];
	walkGlob(root, parts, results);
	return results;
}

function walkGlob(directory: string, parts: string[], results: string[]): void {
	if (parts.length === 0) {
		if (isDirectory(directory)) {
			results.push(directory);
		}
		return;
	}
	const [part, ...rest] = parts;
	if (part === "*") {
		for (const entry of readDirSafe(directory)) {
			const child = `${directory}/${entry}`;
			if (isDirectory(child)) {
				walkGlob(child, rest, results);
			}
		}
		return;
	}
	walkGlob(`${directory}/${part}`, rest, results);
}

function isExcluded(relativePath: string, excludes: string[]): boolean {
	return excludes.some((pattern) =>
		globMatches(relativePath, pattern.replace(/\/$/, "")),
	);
}

function globMatches(value: string, pattern: string): boolean {
	if (pattern === value) {
		return true;
	}
	const regex = new RegExp(`^${pattern.split("*").map(escapeRegex).join(".*")}$`);
	return regex.test(value);
}

function findCurrentMember(cwd: string, members: string[]): string | null {
	const resolvedCwd = resolve(cwd);
	const matches = members.filter((member) => isRelativeTo(resolvedCwd, member));
	if (matches.length === 0) {
		return null;
	}
	return matches.sort((left, right) => right.length - left.length)[0] ?? null;
}

function ancestors(start: string): string[] {
	const result: string[] = [];
	let directory = resolve(start);
	while (true) {
		result.push(directory);
		const parent = resolve(directory, "..");
		if (parent === directory) {
			break;
		}
		directory = parent;
	}
	return result;
}

function isRelativeTo(path: string, parent: string): boolean {
	const normalizedPath = resolve(path);
	const normalizedParent = resolve(parent);
	return (
		normalizedPath === normalizedParent ||
		normalizedPath.startsWith(`${normalizedParent}${sep}`)
	);
}

function isRecord(value: unknown): value is Record<string, unknown> {
	return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isFile(path: string): boolean {
	try {
		return statSync(path).isFile();
	} catch {
		/* v8 ignore next -- covers a race where a path disappears during stat. */
		return false;
	}
}

function isDirectory(path: string): boolean {
	try {
		return statSync(path).isDirectory();
	} catch {
		/* v8 ignore next -- covers a race where a path disappears during stat. */
		return false;
	}
}

function readDirSafe(path: string): string[] {
	try {
		return isDirectory(path) ? readdirSync(path) : [];
	} catch {
		/* v8 ignore next -- covers a race or permissions issue while reading. */
		return [];
	}
}

/* v8 ignore next 2 -- covered through glob matching behavior. */
const escapeRegex = (value: string): string =>
	value.replace(/[.+?^${}()|[\]\\]/g, "\\$&");
